src/ai/backend/storage/xfs/__init__.py

```diff
@@ -60,26 +61,51 @@ async def add_project_entry(
         vfpath = self.backend.mangle_vfpath(vfid)
         if project_id is None:
             project_id = self.get_project_id()
-        await run(
-            [
-                "sudo",
-                "sh",
-                "-c",
-                f"echo '{project_id}:{vfpath}' >> {self.file_projects}",
-            ],
-        )
-        await run(
-            [
-                "sudo",
-                "sh",
-                "-c",
-                f"echo '{str(vfid)}:{project_id}' >> {self.file_projid}",
-            ],
-        )
+
+        temp_name_projects = ""
+        temp_name_projid = ""
+
+        def _create_temp_files():
+            nonlocal temp_name_projects, temp_name_projid
+            try:
+                _projects_content = Path(self.file_projects).read_text()
+                _projects_content += f"{project_id}:{vfpath}\n"
+                _tmp_projects = NamedTemporaryFile(delete=False)
+                _tmp_projects.write(_projects_content.encode("ascii"))
+                temp_name_projects = _tmp_projects.name
+
+                _projid_content = Path(self.file_projid).read_text()
+                _projid_content += f"{str(vfid)}:{project_id}\n"
+                _tmp_projid = NamedTemporaryFile(delete=False)
+                _tmp_projid.write(_projid_content.encode("ascii"))
+                temp_name_projid = _tmp_projid.name
+            finally:
+                if _tmp_projects:
+                    _tmp_projects.close()
+                if _tmp_projid:
+                    _tmp_projid.close()
```

Comment:
Just FYI. (This is not a change request but just a comment.)

If we use Python 3.10 or later, this try-finally block could be rewritten using parenthesized `with` statement and `contextlib.closing()`:
```python3
with (
    closing(NamedTemporaryFile(delete=False)) as _tmp_projects,
    closing(NamedTemporaryFile(delete=False)) as _tmp_projid,
):
    ...
```
Until Python 3.8, this syntax was not allowed due to limitation of the parser and became possible with a new parser in Python 3.9.
Let's update code patterns like this when we fully migrate to Python 3.10. :wink: